### ISSUE TITLE
Issue with AztecOO tolerance

### ISSUE THREAD

**Issue by bartgol** [SNLComputation/Albany#476]
I ran across an annoying fact while working on PR #454 . I'll try to summarize it.

The different solvers-related packages/classes in Trilinos (NOX, Piro, Stratimikos,...) all have their different way of dealing with parameter lists and default values. And the one that started giving me headaches is the linear solvers tolerance when using AztecOO. In particular, the Piro::Epetra classes have different behavior from the Piro ones. In other words, the way you need to specify the tolerance in your input file is different if you use the Piro::Epetra stack or the Piro one, at least when it comes to NOX solver.

Here's what happened to me:
1) used to how easy life was in the Epetra world, I set the AztecOO solver tolerance as (I'm skipping all the other parameters.

```
NOX:
  Direction:
     Newton:
       Stratimikos Linear Solver:
         Stratimikos:
           Linear Solver Types:
             AztecOO:
               Forward Solve:
                 Tolerance: 1e-10
```
but AztecOO was converging to `1e-4`, which is what got me started in my journey.
2) I found out NOX specifies a tolerance for the solver in the 'solveCriteria' variable (which is what Thyra uses to pass solver stopping criteria to the solver). And this tolerance is grabbed from another place in the parameter list. In particular, to get the 1e-10 tolerance, you need

```
NOX:
  Direction:
     Newton:
       Linear Solver:
         Tolerance: 1e-10
```
and whatever you have in the AztecOO sublist will be ignored. Turns out, 1e-4 is the default that NOX uses (well, Albany also sets that value in the sublist, in SolverFactory, so technically we get 1e-4 because of Albany, but the point is that NOX will just use a value for you, without telling you). Not very nice, especially since Belos does not appear to have this problem. You think "whatever, move Tolerance to the 'Linear Solver' sublist". Well, it's not over yet.
3) I removed the `Tolerance` entry in the AztecOO sublist, since it was ignored by NOX. However, the adjoint solves were solved to yet another tolerance, namely 1e-6, which was nowhere to be found in my input file, or in Albany. After hours of digging, I found out that `Thyra::AztecOOLinearOpWithSolve` from Stratimikos was to blame: it sets the tolerance of the adjoint solve to that value. Yes, there should be a way to specify your own in the input file, as
```
NOX:
  Direction:
     Newton:
       Stratimikos Linear Solver:
         Stratimikos:
           Linear Solver Types:
             AztecOO:
               Adjoint Solve:
                 Tolerance: 1e-10
```
but it seems to be ignored. However, if a tolerance is found in the `Forward Solve` list, then that is used.

So here you are, specifying something like this

```
NOX:
  Direction:
     Newton:
       Linear Solver:
         Tolerance: 1e-10
      Stratimikos Linear Solver:
         Stratimikos:
           Linear Solver Types:
             AztecOO:
               Forward Solve:
                 Tolerance: 1e-8
```

which is baloney, since it's very hard to parse: which tolerance is used for the forward linear solves, the one in `Linear Solver` or the une in `Forward Solve`? Also, which of the two is going to be used for the adjoints?

If you are still reading, first, thank you. Second, what should we do? I see three possibilities:

1) Leave this mess as it is, and deal with the nonsense in the input files
2) Make Albany parse input files, and fix it up for Trilinos, so that we can specify tolerances like this:

```
NOX:
  Direction:
     Newton:
      Stratimikos Linear Solver:
         Stratimikos:
           Linear Solver Types:
             AztecOO:
               Forward Solve:
                 Tolerance: 1e-10
               Adjoint Solve:
                 Tolerance: 1e-8
```
Albany would then put the fwd solve tolerance in the `Linear Solver` sublist of `Newton`, so that NOX does not mess up.

3) Get this mess cleaned up in Trilinos (which would likely mean us anyways).

Thoughts?

**Comment by ikalash** [comment]
I’ve been burned by this many times.  It would be nice to get this cleaned up.  If you use belos the linear solver tolerance in the belos list gets ignored.  You have to set it under newton->linear solver.

**Comment by mperego** [comment]
To be clear, this happens because we are no longer using the Piro::Epetra classes, even when using AztecOO, right?
Long term, I think we should fix this in Trilinos. But what would be the desired behaviour? I don't think that going back to the Epetra-like behavior is the right think to do. As an example, we might want different tolerances depending on whether we are solving the state equation or the adjoint equation. 
Actually, I think we should use different solvers for the state equation and for the adjoint equation. @rppawlo can you weigh in on this?

Anyway, for a short term fix I would go with solution 1.

**Comment by rppawlo** [comment]
Yes - this has been a really big pain. Would love to clean it up in Trilinos but I'm not sure what the best path is. 

NOX has algorithms that can change the linear solver tolerance on successive linear solves within a nonlinear solve. This tolerance is based on progress of the nonlinear problem so that you are not over-solving the system when far away from the solution. This means nox needs to know extra info like the initial and final tolerance when using adaptive forcing term capability. It was painful to tell users that if the forcing term is constant then set it in stratimikos or the underlying solver list, but for adaptive, set it in nox. That's why nox has a tolerance parameter. 

So tolerance can be set in:

1. Individual solver list (aztec or belos)
2. Stratimikos
3. NOX

It's hard to remove them at any level. Solvers can be used without stratimikos. Stratimikos can be used without nox.

Also, since users can provide their own nox group (i.e. their own linear solver and linear algebra objects), they can bypass stratimikos altogether, so nox can't even reliably hard code to the stratimikos parameter list.

Last week a new capability was added to teuchos ParameterList that allows for cross-list validation. Basically parameters in one sublist can be dependent on parameters in another sublist. We could change nox get it's final tolerance from the stratimikos list if not set. This is similar to your option 2. Just have albany set some of these new validators. That's a quick fix.

What would be a better option would be to add a query to the NOX Group/LinearOpWithSolve interface that returns the default linear solve tolerance. This change would have to be pushed through the thyra layers and down to all the solvers as well - Belos and AztecOO would need to return their default tolerances. We could use the lower level value as the default in the higher level driver if the parameter is not set at the higher level plist. This is not very satisfying but may be what we need to do since things are split into components. It does force lower level objects to be constructed enough to query parameter before the higher level objects are constructed. I'm not sure I like that either. If you have any design suggestions, I'd love to hear them.

**Comment by bartgol** [comment]
Hi guys, thanks for sharing your thoughts, this is not an easy-solution kind of issue.

@mperego , yes and no, meaning that the Piro::Epetra stack does not seem to exibit this behavior (meaning the tolerances are always grabbed from the linear solver). And AztecOO (within Stratimikos) offers the option of specifying different setting for `Forward Solve` and `Adjoint Solve`. Although, as I said, it _seems_ the `Adjoint Solve` list is not picked up (in which case, the `Forward Solve` parameters are used).

@ikalash you're right. I thought I saw Belos working as expected (i.e., Belos' tolerance being honored), but I just did a second check, and its tolerance is ignored, in favor of the `Linear Solver->Tolerance` one. Just like AztecOO. In a way, that's good, cause it means the same logic works for all solvers. Silver lining.

@rppawlo I don't have a solution either. Perhaps, for something like Albany, it would make sense to have our own way of specifying parameter lists, then have an in-house routine that from the 'albany-style' PL builds a NOX (or whatever) PL. It would not be easy anyways, since we have to know exactly what Trilinos does (and, as this issue suggests, that's not always 100% clear), to be able to fall again in default behaviors.
Perhaps a good starting point could be not let Trilinos packages set defaults for _crucial_ parameters such as the tolerance and the maximum number of iterations. It is unlikely that the user "doesn't care" about the tolerance. If he/she does not specify one (or, perhaps more interestingly, he/she specifies it in the wrong place), Trilinos packages should just throw, with an error saying where the tolerance should instead be specified. This would serve two purposes: 1) avoid undesired defaults to sneak in unnoticed, and 2) teach the user where he/she should specify the parameters. For instance, I could do a dry run without any tolerance/maxiter in my input file, and let trilinos tell me where these parameters were expected but not found.

I understand that removing defaults may cause a massive downpour of breaks in downstream apps, so perhaps this is not a viable path for Trilnios. However, at the very least we could have Trilinos packages issue warnings if a default is used, so that the user has some sort of hope of catching the issue. I could still do a dry run without tolerance/maxiter, and look for all the warnings from trilinos to understand where to patch my input file... How reasonable/doable would this approach be?

**Comment by bartgol** [comment]
Closing, since this seems to be just the way things are in Trilinos. As I said, we could come up with an Albany way of passing parameters, but honestly speaking, we probably won't find the time for this.

**Comment by mperego** [comment]
Update, it seems that things got fixed recently:
```
NOX:
  Direction:
     Method: Newton
     Newton:
       Forcing Term Method: Constant
       Linear Solver:
         Tolerance: 1.0e-07
      Stratimikos Linear Solver:
         Stratimikos:
            Linear Solver Type: Belos
            Linear Solver Types:
              Belos:
                Solver Type: Block GMRES
                Solver Types:
                  Block GMRES:
                    Convergence Tolerance: 1.0e-08
```
Now the tolerance of `1.0e-07` under `Newton/LinearSolver` is used for the linear solvers at each Newton iteration. `Belos` tolerance of `1.0e-08` is only used for the adjoint solve (which does not use NOX). 
Note that if one sets the `Newton/Forcing Term Method` to `Type 1` or `Type 2` instead of `Constant`, then the tolerance of the linear systems are adaptively selected by NOX.
